### ISSUE TITLE
Question: Efficient Handling of `attention_mask` for Long Sequences

### ISSUE THREAD

**Issue by EricLina** [SmallDoges/flash-dmattn#115]
Thank you for your work!

In the documentation, the `attention_mask` is described as an optional mask of shape `(batch_size, 1, query_len, key_len)`. However, this still appears to require initializing a `[L, L]` matrix for the attention mask, where `L` is the sequence length.

**My question:**  
How is this problem addressed when the sequence length `L` is very large? Are there any optimizations or alternative approaches that prevent the memory and computational overhead of allocating such a large attention mask?

Looking forward to your insights!

**Comment by EricLina** [comment]
I think it has nothing to do with qk norm, the mask pattern should be a generation rule like flexattention, otherwise you have to hold a `[L, L]` tensor.

**Comment by LoserCheems** [comment]
However, the flex attention mode seems to be difficult to support multiple indexing of a tensor that requires gradient calculations.

**Comment by EricLina** [comment]
[Vmoba](https://github.com/KwaiVGI/VMoBA/blob/main/src/vmoba.py) calculates with dynamic mask and it is trainable, but it doesn't hold a full `L,L` tensor.

**Comment by LoserCheems** [comment]
Yes, I observed that Vmoba explicitly mentioned using **threshold judgment** for the sparse processing.

**Comment by EricLina** [comment]
Yes, actually it doesn't explicitly pass in a complete mask either.

**Comment by LoserCheems** [comment]
This does not mean that we can make the sampling states bounded and define a threshold to avoid initializing the full mask.

**Comment by weixuansun** [comment]
I am also a little confused on this issue, could you pls provide a sample module code of how to generate attention mask and attention mask from values?

**Comment by EricLina** [comment]
Currently, keeping a large `[L, L]` tensor in memory makes `dmattn` incompatible with very long sequences. For example, if `L=128K`, allocating a `[128K, 128K]` tensor is not feasible due to memory constraints.
I'm curious how you managed experiments with 128K context in your paper. Did you have to store a full `128K x 128K` tensor in memory, or did you use some workaround or optimization to reduce memory usage?

**Comment by LoserCheems** [comment]
As for my testing hardware, for the training scenario, I only tested up to 32K and did not test 128K.

The inference scenario can test longer sequences `[q_len=1, k_len=128K]`, you can try running `benchmarks/forward_performence.py` to obtain the results.

**Comment by LoserCheems** [comment]
@weixuansun The initial idea was to integrate the TopK operation into the CUDA kernel as well.
However, after completing the complex sorting and global reduction for the entire k_len, I found that the performance was not good.

Then, I moved the preprocessing of the mask to the Python side. This not only improved the performance bug also allowed other users to try any mask without having to modify the complex CUDA code.

**Comment by EricLina** [comment]
> In my understanding, the [L, L] attention mask doesn't necessarily be materialized, since it is expanded only from values vector [L, d], nothing to do with q-k similarities. If I understand correctly, it only locate important positions of inputs judged by model weight? I assume that dt_states could be directly fed into attention operation and choose topk indices in a block-wise manner inside the kernel? [@LoserCheems](https://github.com/LoserCheems)

If every row of the mask matrix is predicted in the same way using v, all the rows in the mask would end up being identical. In that sense, I feel it may not truly be "dynamic".

**Comment by EricLina** [comment]
> [@weixuansun](https://github.com/weixuansun) The initial idea was to integrate the TopK operation into the CUDA kernel as well. However, after completing the complex sorting and global reduction for the entire k_len, I found that the performance was not good.
> 
> Then, I moved the preprocessing of the mask to the Python side. This not only improved the performance bug also allowed other users to try any mask without having to modify the complex CUDA code.

One of the main advantages of FlashAttention is that its memory usage is O(N). If you use such a large mask, it will become O(N²).

**Comment by EricLina** [comment]
So your method samples a fixed number of key-value pairs for all attention, right? If that's the case, why do you still need to pass in the full L*L (the complete attention matrix)? Just like @weixuansun mentioned.

**Comment by LoserCheems** [comment]
> > [@weixuansun](https://github.com/weixuansun) The initial idea was to integrate the TopK operation into the CUDA kernel as well. However, after completing the complex sorting and global reduction for the entire k_len, I found that the performance was not good.
> > Then, I moved the preprocessing of the mask to the Python side. This not only improved the performance bug also allowed other users to try any mask without having to modify the complex CUDA code.
> 
> One of the main advantages of FlashAttention is that its memory usage is O(N). If you use such a large mask, it will become O(N²).

However, for the current implementation of the CUDA kernel, there is no need to worry too much about memory issues, because the mask is still in the shape of `kBLockM, kBlockN`, and is chunked and loaded into SRAM. The cost is merely that the size of `kBlock` has been reduced based on the flash-attention.

**Comment by LoserCheems** [comment]
> So your method samples a fixed number of key-value pairs for all attention, right? If that's the case, why do you still need to pass in the full L*L (the complete attention matrix)? Just like [@weixuansun](https://github.com/weixuansun) mentioned.

Because I connected some tests and found that directly passing in the mask and handing the tokens that need to be skipped on the CUDA side is actually faster than the varlen variant's preprocessing of the sequence on the Python side

**Comment by LoserCheems** [comment]
> As you mentioned, the expected shape for `attn_mask` is `(batch_size, nheads, seqlen_q, seqlen_k)`. ([reference](https://github.com/SmallDoges/flash-dmattn/blob/f567c1f14316974e9270b1ec70657dc43c3ff5ac/flash_dmattn/flash_dmattn_interface.py#L1162))
> 
> However, passing such a large mask into your function is not efficient. When attempting to generate an attention mask with the following code:
> 
> B = 1
> H = 12
> L = 32000
> torch.rand([B, H, L, L]).to(torch.bool).to('cuda')
> this step immediately leads to an out-of-memory (OOM) error on CUDA devices.

Indeed, the only solution available at present is to use the varlen function. This allows for ignoring the mask and only requires that the tokens to be calculated be manually packaged on the Python side.

**Comment by LoserCheems** [comment]
> > > > [@weixuansun](https://github.com/weixuansun) The initial idea was to integrate the TopK operation into the CUDA kernel as well. However, after completing the complex sorting and global reduction for the entire k_len, I found that the performance was not good.
> > > > Then, I moved the preprocessing of the mask to the Python side. This not only improved the performance bug also allowed other users to try any mask without having to modify the complex CUDA code.
> > > 
> > > 
> > > One of the main advantages of FlashAttention is that its memory usage is O(N). If you use such a large mask, it will become O(N²).
> > 
> > 
> > However, for the current implementation of the CUDA kernel, there is no need to worry too much about memory issues, because the mask is still in the shape of , and is chunked and loaded into SRAM. The cost is merely that the size of  has been reduced based on the flash-attention.`kBLockM, kBlockN``kBlock`
> 
> I'm not sure if I misunderstood. Although you avoid explicitly materializing the full L*L attention matrix from Q @ K^T by online block calculation in the SRAM, you're still maintaining an extremely large mask in memory.

Indeed, this is a major limitation in the current kernel implementation.

**Comment by EricLina** [comment]
I think passing in a [L, L] mask is not practical, as it can cause OOM (out-of-memory) issues for cases where L > 16k or for larger values.

**Comment by EricLina** [comment]
Another risk issue is that the minimal token granularity is 16 (based on the cuda unit), we couldn't achieve truly dynamic token-level mask.

**Comment by LoserCheems** [comment]
> https://github.com/deepseek-ai/FlashMLA/blob/1408756a88e52a25196b759eaf8db89d2b51b5a1/csrc/sm90/prefill/sparse/fwd.cu [@LoserCheems](https://github.com/LoserCheems) Hello, DeepSeek V3.2 has released their code for dense topk loading kernel, which is a more efficient way than dense mask + scatter.

Great! this is helpful for us!!!

**Comment by EricLina** [comment]
I have a question about their Indexer implemntation, which has to instantiate a score tensor `[N, N]`, and then use torch topk outside of the indexer kernel.

Do you know any optimization methods to fused `score` and `topK` in one kernel with a memeory effient style?
